File: 3_mat_mult_and_pi/pi.cc
#include <stdio.h>
#include <stdlib.h>
#include <iomanip>
#include <omp.h>
#include <cmath>
#include <iostream>
#include <fstream>
#include <sstream>

using namespace std;
double f(double y) {return(4.0/(1.0+y*y));}
int main(int argc, char** argv)
{
double w, x, sum, pi;
int i;
long long n = atoll(argv[1]);
w = 1.0/n;
int nthrd = atoll(argv[2]);
sum = 0.0;


double start_time = omp_get_wtime();

omp_set_num_threads(nthrd);
#pragma omp parallel for private(x) shared(w)\
reduction(+:sum)
for(i=0; i < n; i++)
{
x = w*(i-0.5);
sum = sum + f(x);
}
pi = w*sum;
// printf("pi = %f\n", pi);
cout<<" Number of threads: "<< omp_get_max_threads()<<" Max number of proc.: "<<omp_get_num_procs()<<endl;
cout.precision(10);
double end_time = omp_get_wtime()-start_time;
cout << " Value = "  << pi<< endl;
cout << " Error = " << pi-3.1415926535 << endl;
cout <<" Time = "<<end_time << endl;
}

File: 3_mat_mult_and_pi/pi.f
        program compute_pi
         Implicit None
         include 'omp_lib.h'


         Integer i, argc, nThrd, procs
         Integer*8 n
         Double precision w, x, sum, pi, f, y, truePI, time1, time2
         Character argv1*32
         Character argv2*32
         Integer maxThrd
         Parameter(truePI = 3.1415926535)

         argc = command_argument_count()
         call get_command_argument(1, argv1)
         call get_command_argument(2, argv2)

         READ(argv1, *) nThrd
         READ(argv2, *) n
C       Parameter (n=10000000)
C       Parameter (nThrd=8)
         maxThrd = OMP_GET_MAX_THREADS()
         procs = OMP_GET_NUM_PROCS()

         w = 1.d0/n
         sum = 0.0d0
         time1 = omp_get_wtime() 
!$omp parallel do private(x) num_threads(nThrd)
!$omp& reduction(+:sum)
         do i=1,n
         x = w*(i-0.5d0)
         f = 4.d0/(1.d0 + x*x)
         sum = sum + f
         end do

         pi = w*sum
         time2 = omp_get_wtime()
         Write (6,*) 'Value = ', pi
         Write (6,*) 'Error = ', pi-truePI
         Write (6,*) 'Time = ', time2-time1
         end
